jbmc/unit/java_bytecode/java_object_factory/struct_tag_types.cpp

```diff
@@ -0,0 +1,204 @@
+/*******************************************************************\
+
+Module: Unit tests for Java-object-factory's treatment of tag types
+
+Author: Diffblue Ltd.
+
+\*******************************************************************/
+
+#include <java-testing-utils/load_java_class.h>
+#include <java_bytecode/java_bytecode_language.h>
+#include <java_bytecode/java_object_factory.h>
+#include <langapi/mode.h>
+#include <testing-utils/use_catch.h>
+
+static bool is_decl_with_id(const exprt &expr, const irep_idt &id)
+{
+  if(expr.id() != ID_code || to_code(expr).get_statement() != ID_decl)
+    return false;
+
+  const typet &decl_type = to_code_decl(to_code(expr)).symbol().type();
+  return decl_type.id() == ID_struct_tag &&
+         to_struct_tag_type(decl_type).get_identifier() == id;
+}
+
+static bool is_decl_of_type(const exprt &expr, const typet &type)
+{
+  return expr.id() == ID_code && to_code(expr).get_statement() == ID_decl &&
+         to_code_decl(to_code(expr)).symbol().type() == type;
+}
+
+static bool
+contains(const exprt &expr, std::function<bool(const exprt &)> predicate)
+{
+  if(predicate(expr))
+    return true;
+  for(const auto &op : expr.operands())
+  {
+    if(contains(op, predicate))
+      return true;
+  }
+  return false;
+}
+
+static bool
+contains(const typet &type, std::function<bool(const typet &)> predicate)
+{
+  if(predicate(type))
+    return true;
+
+  if(type.has_subtypes())
+  {
+    for(const auto &subtype : to_type_with_subtypes(type).subtypes())
+    {
+      if(contains(subtype, predicate))
+        return true;
+    }
+  }
+
+  return false;
+}
+
+static bool contains_decl_of_type(const exprt &expr, const typet &type)
+{
+  return contains(expr, [&type](const exprt &subexpr) {
```

Comment:
:pick: Here you do a full tree search of the entire ast for code_decl, but in practise it is only `code_blockt` you need to flatten? `require_goto_statements::get_all_statements` does the top level flattening. This has the advantage of not being misleading since a decl can't really appear anywhere in the AST.